pkg/elle/txn/cycle.go

```diff
@@ -0,0 +1,214 @@
+package txn
+
+import (
+	"strings"
+
+	"github.com/pingcap/tipocket/pkg/elle/core"
+)
+
+type FilterExType = func(cycleCase *CycleCase) bool
+
+type CycleAnomalySpecType struct {
+	// A set of relationships which must intersect with every edge in the cycle.
+	Rels map[core.Rel]struct{}
+
+	// A set of relationships which must intersect with the first edge in the cycle.
+	FirstRel core.Rel
+	// A set of relationships which must intersect with remaining edges.
+	RestRels map[core.Rel]struct{}
+
+	// A predicate over a cycle explanation. We use this to restrict cycles to e.g. *just* G2 instead of G-single.
+	FilterEx FilterExType
+
+	With            func(n int, lastIsRw bool, rel core.Rel) (bool, int, bool) // seems only G-nonadjacent is used, it's a little complicated
+	FilterPathState func(interface{}) bool
+}
+
+var CycleAnomalySpecs map[string]CycleAnomalySpecType
+var CycleTypeNames map[string]struct{}
+var UnknownAnomalyTypes map[string]struct{}
+
+// Anomaly types involving realtime edges.
+var RealtimeAnalysisTypes map[string]struct{}
+
+var ProcessAnalysisTypes map[string]struct{}
+
+func fromRels(rels ...core.Rel) CycleAnomalySpecType {
+	return fromRelsWithFilter(nil, rels...)
+}
+
+func fromRelsWithFilter(filter FilterExType, rels ...core.Rel) CycleAnomalySpecType {
+	relsSet := map[core.Rel]struct{}{}
+	for _, v := range rels {
+		relsSet[v] = struct{}{}
+	}
+	return CycleAnomalySpecType{
+		Rels:     relsSet,
+		FilterEx: filter,
+	}
+}
+
+func fromFirstRelAndRest(first core.Rel, rests ...core.Rel) CycleAnomalySpecType {
+	return fromFirstRelAndRestWithFilter(nil, first, rests...)
+}
+
+func fromFirstRelAndRestWithFilter(filter FilterExType, first core.Rel, rests ...core.Rel) CycleAnomalySpecType {
+	restSet := map[core.Rel]struct{}{}
+	for _, v := range rests {
+		restSet[v] = struct{}{}
+	}
+	return CycleAnomalySpecType{
+		FirstRel: first,
+		RestRels: restSet,
+		FilterEx: filter,
+	}
+}
+
+func filterClosure(required string) FilterExType {
```

Comment:
Maybe buildFilterExType or mkFilterExType is better.

Reply:
Good idea.